File: scripts/fetch_species_photos.py
"""Fetch species photos from iNaturalist API for species missing images."""
import json
import time
import urllib.request
import urllib.parse
import ssl

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

PHOTOS_FILE = "frontend/public/data/species-photos.json"
APP_DATA_FILE = "frontend/public/data/app-data.json"

with open(APP_DATA_FILE) as f:
    app_data = json.load(f)
with open(PHOTOS_FILE) as f:
    photos = json.load(f)

species_ids = [n["id"] for n in app_data.get("nodes", [])]
missing = [s for s in species_ids if s not in photos]
print(f"{len(missing)} species missing photos", flush=True)

found = 0
failed = []

for i, name in enumerate(missing):
    try:
        url = f"https://api.inaturalist.org/v1/taxa?q={urllib.parse.quote(name)}&per_page=5"
        req = urllib.request.Request(url, headers={"User-Agent": "BioScope/1.0"})
        with urllib.request.urlopen(req, timeout=10, context=ctx) as resp:
            data = json.loads(resp.read())

        photo_url = None
        for result in data.get("results", []):
            if result.get("name", "").lower() == name.lower():
                dp = result.get("default_photo")
                if dp and dp.get("medium_url"):
                    photo_url = dp["medium_url"]
                    break

        if not photo_url and data.get("results"):
            dp = data["results"][0].get("default_photo")
            if dp and dp.get("medium_url"):
                photo_url = dp["medium_url"]

        if photo_url:
            photos[name] = photo_url
            found += 1
        else:
            failed.append(name)

        if (i + 1) % 25 == 0:
            print(f"  {i+1}/{len(missing)} processed, {found} found", flush=True)
            with open(PHOTOS_FILE, "w") as f:
                json.dump(photos, f, indent=2)

        time.sleep(0.5)

    except Exception as e:
        failed.append(name)
        if (i + 1) % 25 == 0:
            print(f"  {i+1}/{len(missing)} processed, {found} found (error: {e})", flush=True)

with open(PHOTOS_FILE, "w") as f:
    json.dump(photos, f, indent=2)

print(f"\nDone! Found {found}/{len(missing)} photos. {len(failed)} failed.", flush=True)
if failed:
    print(f"Failed species: {failed[:10]}{'...' if len(failed) > 10 else ''}", flush=True)

File: scripts/fetch_species_photos.js
const fs = require('fs');
const https = require('https');
const path = require('path');

const MASTER_SPECIES_PATH = 'frontend/src/lib/masterSpecies.json';
const OUTPUT_PATH = 'frontend/public/data/species-photos.json';

if (!fs.existsSync(MASTER_SPECIES_PATH)) {
  console.error('masterSpecies.json not found. Run process_data.py first.');
  process.exit(1);
}

const masterSpecies = JSON.parse(fs.readFileSync(MASTER_SPECIES_PATH, 'utf8'));
const speciesNames = Object.keys(masterSpecies);

function fetchJSON(url, retryCount = 0) {
  return new Promise((resolve, reject) => {
    https.get(url, { 
      headers: { 'User-Agent': 'BioScope/1.0' },
      timeout: 10000 
    }, (res) => {
      if (res.statusCode === 429) {
        return reject({ code: 429, message: 'Too Many Requests' });
      }
      
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try { 
          const parsed = JSON.parse(data);
          resolve(parsed); 
        } catch (e) { 
          reject(new Error(`Parse error: ${data.slice(0, 100)}...`)); 
        }
      });
    }).on('error', (err) => {
      reject(err);
    }).on('timeout', () => {
      reject(new Error('Request Timeout'));
    });
  });
}

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

async function main() {
  let photos = {};
  if (fs.existsSync(OUTPUT_PATH)) {
    try { photos = JSON.parse(fs.readFileSync(OUTPUT_PATH, 'utf8')); }
    catch (e) { console.warn('Could not read existing photos.'); }
  }

  let found = 0;
  let skipped = 0;
  let errors = 0;

  console.log(`Expansion starting: 0/${speciesNames.length} species checked`);

  for (let i = 0; i < speciesNames.length; i++) {
    const name = speciesNames[i];
    
    if (photos[name] !== undefined && photos[name] !== null) {
      skipped++;
      found++;
      continue;
    }

    const url = `https://api.inaturalist.org/v1/taxa?q=${encodeURIComponent(name)}&per_page=1`;
    let attempts = 0;
    let success = false;

    while (attempts < 3 && !success) {
      try {
        const data = await fetchJSON(url);
        success = true;
        if (data.results && data.results.length > 0) {
          const taxon = data.results[0];
          if (taxon.default_photo && (taxon.default_photo.square_url || taxon.default_photo.url)) {
            let photoUrl = taxon.default_photo.medium_url || taxon.default_photo.url || taxon.default_photo.square_url;
            photoUrl = photoUrl.replace('/square', '/medium'); // Upgrade if needed
            photos[name] = photoUrl;
            found++;
            console.log(`[${i + 1}/${speciesNames.length}] ${name} -> OK`);
          } else {
            photos[name] = null;
          }
        } else {
          photos[name] = null;
        }
      } catch (e) {
        attempts++;
        if (e.code === 429 || attempts < 3) {
          const wait = e.code === 429 ? 5000 * attempts : 1000 * attempts;
          console.warn(`[${i + 1}/${speciesNames.length}] ${name} -> Retrying (${attempts}/3) in ${wait}ms...`);
          await sleep(wait);
        } else {
          console.error(`[${i + 1}/${speciesNames.length}] ${name} -> Final Error: ${e.message}`);
          errors++;
        }
      }
    }

    // Moderate rate limit for success
    await sleep(200);
    
    if (i % 10 === 0) {
      fs.writeFileSync(OUTPUT_PATH, JSON.stringify(photos, null, 2));
    }
  }

  fs.writeFileSync(OUTPUT_PATH, JSON.stringify(photos, null, 2));
  console.log(`\nExpansion Complete:`);
  console.log(`  - Total: ${speciesNames.length}`);
  console.log(`  - Found: ${found}`);
  console.log(`  - Skipped (Cache): ${skipped}`);
  console.log(`  - Errors: ${errors}`);
}

main().catch(console.error);
